contracts/BalanceTrackerBase.sol

```diff
@@ -196,95 +204,129 @@ abstract contract BalanceTrackerBase {
+                if (requesterDeliveryRates[i] > mechDeliveryRates[i]) {
+                    // Return back requester overpayment debit / credit
+                    rateDiff = requesterDeliveryRates[i] - mechDeliveryRates[i];
+                    totalRateDiff += rateDiff;
+
+                    // Adjust requester balance
+                    balance = _adjustFinalBalance(requesters[i], rateDiff);
+                    mapRequesterBalances[requesters[i]] = balance;
+                }
+            }
         }
 
-        // Check for delivery rate difference
-        uint256 rateDiff;
-        uint256 balance;
-        if (maxDeliveryRate > actualDeliveryRate) {
-            // Return back requester overpayment debit / credit
-            rateDiff = maxDeliveryRate - actualDeliveryRate;
-
-            // Adjust requester balance
-            balance = _adjustFinalBalance(requester, rateDiff);
-            mapRequesterBalances[requester] = balance;
-        } else {
-            // Limit the rate by the max chosen one as that is what the requester agreed on
-            actualDeliveryRate = maxDeliveryRate;
+        // Check for zero value
+        if (totalMechDeliveryRate == 0) {
+            revert ZeroValue();
         }
 
         // Record payment into mech balance
         balance = mapMechBalances[mech];
-        balance += actualDeliveryRate;
+        balance += totalMechDeliveryRate;
         mapMechBalances[mech] = balance;
 
-        emit MechPaymentCalculated(mech, requestId, actualDeliveryRate, rateDiff);
-        emit MechBalanceAdjusted(mech, actualDeliveryRate, balance);
-
-        _locked = 1;
+        emit MechBalanceAdjusted(mech, totalMechDeliveryRate, balance, totalRateDiff);
     }
 
-    /// @dev Adjusts requester and mech balances for direct batch request processing.
+    /// @dev Adjusts mech and requester balances for direct batch request processing.
     /// @notice This function can be called by the Mech Marketplace only.
     /// @param mech Mech address.
     /// @param requester Requester address.
-    /// @param totalDeliveryRate Total batch delivery rate.
-    /// @param paymentData Additional payment-related request data, if applicable.
-    function adjustRequesterMechBalances(
+    /// @param mechDeliveryRates Set of actual charged delivery rates for each request.
+    function adjustMechRequesterBalances(
         address mech,
         address requester,
-        uint256 totalDeliveryRate,
-        bytes memory paymentData
-    ) external payable {
-        checkAndRecordDeliveryRate(requester, totalDeliveryRate, paymentData);
+        uint256[] memory mechDeliveryRates,
+        bytes memory
+    ) external virtual {
+        // Reentrancy guard
+        if (locked) {
+            revert ReentrancyGuard();
+        }
+        locked = true;
+
+        // Get total mech delivery rate
+        uint256 totalMechDeliveryRate;
+        for (uint256 i = 0; i < mechDeliveryRates.length; ++i) {
```

Comment:
This is now simplified for requests with signatures: simply calculate all the mech prices, take from requester and give to a delivery mech.